doc/design.md

```diff
@@ -0,0 +1,1268 @@
+Data changes more frequently so copying it is probably feasible.
+However, this certainly has negative effects on memory usage and performance.
+Furthermore it would significantly reduce the flexibility gained with `Dataset` since suddenly having datasets with many variables will become a burden instead of an opportunity.
+
+
+### <a name="design-details-uncertainties-are-stored-as-variances"></a>Uncertainties are stored as variances, not standard deviations
+
+##### Rationale
+
+Mantid currently stores uncertainties as standard deviation (except for `DataObjects::MDHistoWorkspace`).
+However, most computations require the variance for propagation of uncertainties.
+To this end it is often squared in place (breaking assumptions about, e.g., `HistogramData::Histogram` and `DataObjects::Workspace2D`, and subsequently converted back to standard deviation by taking the square root.
+This also adds a little computational overhead.
+Having a clear and well defined strategy will reduce the risk of confusion.
+
+##### Implementation
+
+The go-to tag for storing uncertainties will be `Data::Variance`.
+`Data::StdDev` will exist, but is for *access* only, i.e., attempting to create a `Variable` with that tag should fail.
+Instead `Data::StdDev` can by used in combination with `DatasetView` to *read* the standard deviations if required, e.g., for visualization purposes.
+If it turns out to be useful it would also be possible to provide setters --- as opposed to a simple assignment which is possible if the actual underlying data is referenced --- for standard deviation using `DatasetView` .
+
+
+### <a name="design-details-distribution-histogram-distinguised-by-unit"></a>"Distributions" and "histogram data" are distinguished by their unit
+
+##### Rationale
+
+Currently `API::MatrixWorkspace` (and by extension `HistogramData::Histogram`) keep a flag indicating whether data in a histogram has been divided by its bin width.
+A considerable number of algorithms are checking this manually, branching to different code sections.
+`HistogramData::Histogram` attempted to solve this by introducing getters like `Histogram::counts()` and `Histogram::frequencies()`.
+However, it seems to be useful only in a rather limited number of places, whereas it adds a significant amount of complication to the API.
+The flag adds complication and should thus be removed.
+
+##### Implementation
+
+This will require having a unit for "counts", i.e., we cannot simply use "dimensionless".
+We need to establish whether it is possible to write clean client code based on the unit-driven mechanism.
+Apart from that there should not be any major implementation difficulties.
+
+
+### <a name="design-details-no-progress-reporting-no-cancellation"></a>No progress reporting and no cancellation
+
+##### Rationale
+
+- Most operations are simple and thus fast.
+- Cancellation and progress reporting can be implemented in higher-level algorithms that *use* `Dataset`.
```

Comment:
Note that there will be certain expensive operations that cannot be cancelled, due to having a single multi dimensional array, instead of many one-dimensional ones. In particular, allocating a large output array, e.g., for `rebin` is the most expensive part.

Reply:
👍 on @martyngigg 's recommendation for a callback system. Here's an example where we use VTK's callback system in Mantid.

https://github.com/mantidproject/mantid/blob/21b138884c35d8daaac103c5d7f5c890f6a50c56/qt/paraview_ext/VatesAlgorithms/src/SaveMDWorkspaceToVTKImpl.cpp#L231